### PR TITLE
[autogluon][build] AutoGluon 0.4.0 container

### PR DESCRIPTION
**Pull request by gradientsky**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Description
AutoGluon 0.4.0 containers release

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [x] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [x] `sagemaker_local_tests = true`

## Label Checklist
- [x] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [x] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [x] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [x] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: autogluon/training/docker/0.4.0/py3/cu112/Dockerfile.gpu [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+ARG FROM_IMAGE=rapidsai/rapidsai-core
+
+FROM ${FROM_IMAGE}:cuda${CUDA_VER}-base-${LINUX_VER}-py${PYTHON_VER}
+
+LABEL maintainer="Amazon AI"
+LABEL dlc_major_version="1"
+
+ARG CUDA_VER
+
+ARG OPENSSL_VERSION=1.1.1l
```
Comment (saimidu) [inline autogluon/training/docker/0.4.0/py3/cu112/Dockerfile.gpu:14]: Please update this to the latest openssl version (I think `1.1.1n`).

File: autogluon/training/docker/0.4.0/py3/cu112/Dockerfile.gpu [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+RUN wget -q -c https://www.openssl.org/source/openssl-${OPENSSL_VERSION}.tar.gz \
+ && tar -xzf openssl-${OPENSSL_VERSION}.tar.gz \
+ && cd openssl-${OPENSSL_VERSION} \
+ && ./config && make -j $(nproc) && make install \
+ && ldconfig \
+ && cd .. && rm -rf openssl-* \
+ && rmdir /usr/local/ssl/certs \
+ && ln -s /etc/ssl/certs /usr/local/ssl/certs
+
+ARG CUDNN_VERSION=8.*
```
Comment (saimidu) [inline autogluon/training/docker/0.4.0/py3/cu112/Dockerfile.gpu:54]: Please define this at the top of the dockerfile, so that it can be easily found and modified if needed. Perhaps next to `CUDA_VER`?

File: autogluon/training/docker/0.4.0/py3/cu112/Dockerfile.gpu [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+ARG PYTHON_VERSION=3.8.12
+ARG PYTHON_VER=3.8
+ARG FROM_IMAGE=rapidsai/rapidsai-core
+
+FROM ${FROM_IMAGE}:cuda${CUDA_VER}-base-${LINUX_VER}-py${PYTHON_VER}
+
+LABEL maintainer="Amazon AI"
+LABEL dlc_major_version="1"
+
+ARG CUDA_VER
```
Comment (saimidu) [inline autogluon/training/docker/0.4.0/py3/cu112/Dockerfile.gpu:12]: Why is this needed when `ARG CUDA_VER=11.2` is defined at the top of the dockerfile?

File: autogluon/training/docker/0.4.0/py3/cu112/Dockerfile.gpu [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+ && curl -o ${HOME_DIR}/oss_compliance.zip https://aws-dlinfra-utilities.s3.amazonaws.com/oss_compliance.zip \
+ && unzip ${HOME_DIR}/oss_compliance.zip -d ${HOME_DIR}/ \
+ && chmod 775 -R ${HOME_DIR}/oss_compliance/ \
+ && cp ${HOME_DIR}/oss_compliance/test/testOSSCompliance /usr/local/bin/testOSSCompliance \
+ && chmod +x /usr/local/bin/testOSSCompliance \
+ && chmod +x ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh \
+ && ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh ${HOME_DIR} python3 \
+ && rm -rf ${HOME_DIR}/oss_compliance*
+
+RUN curl -o /license.txt https://autogluon.s3.us-west-2.amazonaws.com/licenses/THIRD-PARTY-LICENSES.txt
```
Comment (saimidu) [inline autogluon/training/docker/0.4.0/py3/cu112/Dockerfile.gpu:129]: This link seems broken.

File: autogluon/inference/docker/0.4.0/py3/Dockerfile.cpu [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+ARG AUTOGLUON_VERSION=0.4.0
+
+RUN pip --no-cache-dir install --upgrade --trusted-host pypi.org --trusted-host files.pythonhosted.org pip \
+ && pip --no-cache-dir install --upgrade wheel setuptools \
+ && pip uninstall -y dataclasses \
+ && pip install --no-cache-dir -U torch==${TORCH_VER} torchvision==${TORCH_VISION_VER} -f https://download.pytorch.org/whl/cpu/torch_stable.html \
+    numpy==${NUMPY_VER} \
+    autogluon==${AUTOGLUON_VERSION}
+
+EXPOSE 8080 8081
```
Comment (saimidu) [inline autogluon/inference/docker/0.4.0/py3/Dockerfile.cpu:33]: The AG `THIRD-PARTY-LICENSES.txt` that is downloaded into the training images seems to be missing in this dockerfile. Is this intentional?

File: autogluon/buildspec.yml [hunk 6]
```diff
@@ -1,8 +1,8 @@
 account_id: &ACCOUNT_ID <set-$ACCOUNT_ID-in-environment>
 region: &REGION <set-$REGION-in-environment>
 framework: &FRAMEWORK autogluon
-version: &VERSION 0.3.2
-shart_version: &SHORT_VERSION 0.3
+version: &VERSION 0.4.0
+shart_version: &SHORT_VERSION 0.4
```
[the remark this answers is not in the document]
Comment (gradientsky) [reply autogluon/buildspec.yml:5]: We can't use short version because AG 0.3.1 is released in 2 different containers 0.3.1 and 0.3.2

[the remark this answers is not in the document]
Comment (gradientsky) [reply autogluon/buildspec.yml:5]: fixed typos

File: autogluon/buildspec_0_3_2.yml [hunk 7]
```diff
@@ -0,0 +1,74 @@
+account_id: &ACCOUNT_ID <set-$ACCOUNT_ID-in-environment>
```
[the remark this answers is not in the document]
Comment (gradientsky) [reply autogluon/buildspec_0_3_2.yml:1]: done

File: autogluon/inference/docker/0.4.0/py3/Dockerfile.cpu [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+ARG PYTHON_VERSION=3.8.10
+
+RUN apt-get update \
+ && apt-get -y upgrade \
+ && apt-get autoremove -y \
+ && apt-get clean \
+ && rm -rf /var/lib/apt/lists/*
+
+ARG TORCH_VER=1.10.0+cpu
```
[the remark this answers is not in the document]
Comment (gradientsky) [reply autogluon/inference/docker/0.4.0/py3/Dockerfile.cpu:21]: We need CPU version and there's no `1.10+cpu` option

File: autogluon/training/docker/0.4.0/py3/Dockerfile.cpu [hunk 9]
```diff
@@ -0,0 +1,173 @@
+ARG FROM_IMAGE=ubuntu:18.04
+FROM ${FROM_IMAGE}
+
+LABEL maintainer="Amazon AI"
+LABEL dlc_major_version="1"
+
+ARG OPENSSL_VERSION=1.1.1l
```
[the remark this answers is not in the document]
Comment (gradientsky) [reply autogluon/training/docker/0.4.0/py3/Dockerfile.cpu:7]: done

File: autogluon/training/docker/0.4.0/py3/Dockerfile.cpu [hunk 10]
```diff
@@ -0,0 +1,173 @@
+ARG FROM_IMAGE=ubuntu:18.04
```
[the remark this answers is not in the document]
Comment (gradientsky) [reply autogluon/training/docker/0.4.0/py3/Dockerfile.cpu:1]: done

File: autogluon/training/docker/0.4.0/py3/cu112/Dockerfile.gpu (hunk 1 above)
Comment (gradientsky) [reply autogluon/training/docker/0.4.0/py3/cu112/Dockerfile.gpu:14]: done

File: autogluon/training/docker/0.4.0/py3/cu112/Dockerfile.gpu (hunk 2 above)
Comment (gradientsky) [reply autogluon/training/docker/0.4.0/py3/cu112/Dockerfile.gpu:54]: done

File: autogluon/training/docker/0.4.0/py3/cu112/Dockerfile.gpu (hunk 4 above)
Comment (gradientsky) [reply autogluon/training/docker/0.4.0/py3/cu112/Dockerfile.gpu:132]: good catch - fixed

File: autogluon/inference/docker/0.4.0/py3/Dockerfile.cpu (hunk 5 above)
Comment (gradientsky) [reply autogluon/inference/docker/0.4.0/py3/Dockerfile.cpu]: added

File: autogluon/training/docker/0.4.0/py3/Dockerfile.cpu [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+RUN HOME_DIR=/root \
+ && curl -o ${HOME_DIR}/oss_compliance.zip https://aws-dlinfra-utilities.s3.amazonaws.com/oss_compliance.zip \
+ && unzip ${HOME_DIR}/oss_compliance.zip -d ${HOME_DIR}/ \
+ && cp ${HOME_DIR}/oss_compliance/test/testOSSCompliance /usr/local/bin/testOSSCompliance \
+ && chmod +x /usr/local/bin/testOSSCompliance \
+ && chmod +x ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh \
+ && ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh ${HOME_DIR} ${PYTHON} \
+ && rm -rf ${HOME_DIR}/oss_compliance*
+
+RUN curl -o /license.txt https://autogluon.s3.us-west-2.amazonaws.com/licenses/THIRD-PARTY-LICENSES.txt
```
[the remark this answers is not in the document]
Comment (gradientsky) [reply autogluon/training/docker/0.4.0/py3/Dockerfile.cpu:177]: fixed

File: autogluon/training/docker/0.4.0/py3/cu112/Dockerfile.gpu [hunk 12]
```diff
@@ -0,0 +1,131 @@
+ARG CUDA_VER=11.2
+ARG LINUX_VER=ubuntu18.04
```
[the remark this answers is not in the document]
Comment (gradientsky) [reply autogluon/training/docker/0.4.0/py3/cu112/Dockerfile.gpu]: fixed

File: autogluon/training/docker/0.4.0/py3/Dockerfile.cpu [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+FROM ${FROM_IMAGE}
+
+LABEL maintainer="Amazon AI"
+LABEL dlc_major_version="1"
+
+ARG OPENSSL_VERSION=1.1.1l
+ARG PYTHON_VERSION=3.8.12
+ARG PYTHON=python3
+ARG MXNET_VER=1.9.*
+ARG TORCH_VER=1.10.0+cpu
```
[the remark this answers is not in the document]
Comment (gradientsky) [reply autogluon/training/docker/0.4.0/py3/Dockerfile.cpu]: ditto (see above)

File: autogluon/buildspec.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (saimidu) [reply autogluon/buildspec.yml]: Let's only do this for AG 0.4. The 0.3.1 and 0.3.2 versions can stay in their folders as they are now.

File: autogluon/inference/docker/0.4.0/py3/Dockerfile.cpu [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+ && pip uninstall -y dataclasses \
+ && pip install --no-cache-dir -U torch==${TORCH_VER} torchvision==${TORCH_VISION_VER} -f https://download.pytorch.org/whl/cpu/torch_stable.html \
+    autogluon==${AUTOGLUON_VERSION}
+
+# Security patching
+RUN pip install --no-cache-dir -U 'ray>=1.10,<1.11' \
+ && pip install --no-cache-dir -U 'pytorch_lightning>=1.5.10,<1.7.0' \
+ && pip install --no-cache-dir -U 'Pillow>=9.0.1,<9.1.0'
+
+RUN curl -o /license.txt https://autogluon.s3.us-west-2.amazonaws.com/licenses/THIRD-PARTY-LICENSES.txt
```
Comment (saimidu) [inline autogluon/inference/docker/0.4.0/py3/Dockerfile.cpu:36]: Could you put this file in its own location? Do we want to overwrite the MXNet 1.9 license already present in the base image?

Comment (gradientsky) [reply autogluon/inference/docker/0.4.0/py3/Dockerfile.cpu]: Updated to write to `license-autogluon.txt`

**Review by saimidu — APPROVED** [review]

**Review by saimidu — APPROVED** [review]
